From 23b2025c0435a842243cf2346b5f284272f2198e Mon Sep 17 00:00:00 2001
From: Stephan Mueller <smueller@chronox.de>
Date: Sat, 7 Jan 2023 17:29:11 +0100
Subject: [PATCH 30/35] lrng: fix return code of RNDADDENTROPY

The legacy /dev/random returns 0 on success for RNDADDENTROPY. The LRNG
returned a positive integer equal to the number of bytes inserted into
the LRNG. To stay ABI compliant, the LRNG returns 0 on success now as
well.

Reported-by: Dominic Esposito
Signed-off-by: Stephan Mueller <smueller@chronox.de>
---
 drivers/char/lrng/lrng_drng_mgr.c             | 2 +-
 drivers/char/lrng/lrng_interface_dev_common.c | 7 ++++---
 2 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/char/lrng/lrng_drng_mgr.c b/drivers/char/lrng/lrng_drng_mgr.c
index 8939f1777..1b9539eec 100644
--- a/drivers/char/lrng/lrng_drng_mgr.c
+++ b/drivers/char/lrng/lrng_drng_mgr.c
@@ -147,7 +147,7 @@ int lrng_drng_alloc_common(struct lrng_drng *drng,
 	drng->drng_cb = drng_cb;
 	drng->drng = drng_cb->drng_alloc(LRNG_DRNG_SECURITY_STRENGTH_BYTES);
 	if (IS_ERR(drng->drng))
-		return PTR_ERR(drng->drng);
+		return -PTR_ERR(drng->drng);
 
 	lrng_drng_reset(drng);
 	return 0;
diff --git a/drivers/char/lrng/lrng_interface_dev_common.c b/drivers/char/lrng/lrng_interface_dev_common.c
index 09d056307..a2845e278 100644
--- a/drivers/char/lrng/lrng_interface_dev_common.c
+++ b/drivers/char/lrng/lrng_interface_dev_common.c
@@ -246,7 +246,7 @@ ssize_t lrng_drng_write(struct file *file, const char __user *buffer,
 long lrng_ioctl(struct file *f, unsigned int cmd, unsigned long arg)
 {
 	u32 digestsize_bits;
-	int size, ent_count_bits;
+	int size, ent_count_bits, ret;
 	int __user *p = (int __user *)arg;
 
 	switch (cmd) {
@@ -281,8 +281,9 @@ long lrng_ioctl(struct file *f, unsigned int cmd, unsigned long arg)
 			return -EINVAL;
 		/* there cannot be more entropy than data */
 		ent_count_bits = min(ent_count_bits, size<<3);
-		return lrng_drng_write_common((const char __user *)p, size,
-					      ent_count_bits);
+		ret = lrng_drng_write_common((const char __user *)p, size,
+					     ent_count_bits);
+		return (ret < 0) ? ret : 0;
 	case RNDZAPENTCNT:
 	case RNDCLEARPOOL:
 		/* Clear the entropy pool counter. */
-- 
2.39.0.rc2.1.gbd5df96b79

